homeassistant/components/switchbot/config_flow.py

```diff
@@ -0,0 +1,187 @@
+"""Config flow for Switchbot."""
+from __future__ import annotations
+
+from asyncio import Lock
+import logging
+from typing import Any
+
+from switchbot import GetSwitchbotDevices  # pylint: disable=import-error
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry, ConfigFlow, OptionsFlow
+from homeassistant.const import CONF_MAC, CONF_NAME, CONF_PASSWORD, CONF_SENSOR_TYPE
+from homeassistant.core import callback
+from homeassistant.data_entry_flow import FlowResult
+
+from .const import (
+    ATTR_BOT,
+    BTLE_LOCK,
+    CONF_RETRY_COUNT,
+    CONF_RETRY_TIMEOUT,
+    CONF_SCAN_TIMEOUT,
+    CONF_TIME_BETWEEN_UPDATE_COMMAND,
+    DEFAULT_RETRY_COUNT,
+    DEFAULT_RETRY_TIMEOUT,
+    DEFAULT_SCAN_TIMEOUT,
+    DEFAULT_TIME_BETWEEN_UPDATE_COMMAND,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _btle_connect(mac: str) -> dict:
+    """Scan for BTLE advertisement data."""
+    # Try to find switchbot mac in nearby devices,
+    # by scanning for btle devices.
+
+    switchbots = GetSwitchbotDevices()
+    switchbots.discover()
+    switchbot_device = switchbots.get_device_data(mac=mac)
+
+    if not switchbot_device:
+        raise NotConnectedError("Failed to discover switchbot")
+
+    return switchbot_device
+
+
+class SwitchbotConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Switchbot."""
+
+    VERSION = 1
+
+    async def _validate_mac(self, data: dict) -> FlowResult:
+        """Try to connect to Switchbot device and create entry if successful."""
+        await self.async_set_unique_id(data[CONF_MAC].replace(":", ""))
+        self._abort_if_unique_id_configured()
+
+        # asyncio.lock prevents btle adapter exceptions if there are multiple calls to this method.
+        # store asyncio.lock in hass data if not present.
+        if not self.hass.data.get(DOMAIN):
+            self.hass.data.setdefault(DOMAIN, {})
+            self.hass.data[DOMAIN][BTLE_LOCK] = Lock()
```

Comment:
```suggestion
        if BTLE_LOCK not in self.hass.data[DOMAIN]:
            self.hass.data[DOMAIN][BTLE_LOCK] = Lock()
```

If there is ever a change in `async_setup_entry` that `await`s after creating the empty. dict in `hass.data[DOMAIN]`, it could return control to the event loop which would mean `hass.data[DOMAIN]` would exist but the BTLE_LOCK had not been created yet. This code would create the lock instead of `async_setup_entry`.  Right now there is a check in `async_setup_entry` to prevent creating it so it should be safe.  However without explicitly checking there is some risk we could clobber the lock here due to later refactoring.